File: leetcode/src/main/java/org/luvx/leetcode/java/medium/_300/Solution.java
package org.luvx.leetcode.java.medium._300;
/*
 * @lc app=leetcode.cn id=300 lang=java
 *
 * [300] 最长递增子序列
 */

import java.util.ArrayList;
import java.util.List;

// @lc code=start
class Solution {
    int result = 1;

    /**
     * dp[i]的结果计算并不依赖dp[i-1]
     */
    public int lengthOfLIS(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        int[] dp = new int[nums.length + 1];
        dp[0] = Integer.MIN_VALUE;
        dp[1] = nums[0];
        int max = 1;
        for (int i = 1; i < nums.length; i++) {
            if (dp[max] < nums[i]) {
                max++;
                dp[max] = nums[i];
                continue;
            }

            for (int j = max - 1; j >= 0; j--) {
                if (nums[i] > dp[j]) {
                    dp[j + 1] = nums[i];
                    break;
                }
            }
        }
        return max;
    }

    /**
     * 动态规划(最容易理解, 但时间复杂度在n^2)
     */
    public int lengthOfLIS0(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        int[] dp = new int[nums.length];
        dp[0] = 1;

        int max = 1;
        for (int i = 1; i < nums.length; i++) {
            dp[i] = 1;
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
            max = Math.max(max, dp[i]);
        }
        return max;
    }

    /**
     * 10, 9, 2, 5, 3, 7, 101, 18
     * tails: [2, 3, 7, 18, 0, 0, 0, 0]
     */
    public int lengthOfLIS2(int[] nums) {
        // 每一项 p[i] 的含义是，所有长度为 i+1 的上升子序列的末尾元素中的最小值
        int[] tails = new int[nums.length];
        tails[0] = nums[0];
        int max = 1; // 因为用于下标, [left,max]所表示的数据范围长度为max+1
        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            // [0, 目前最长的地方] 范围内查找
            int left = 0, right = max;
            // 此处的二分查找, 最终是定位left值(num 在有序 tails 中的位置)
            while (left < right) {
                int mid = left + (right - left) / 2;
                if (tails[mid] < num) {
                    // 进入这里,说明是递增, tails才会增加元素, 子序列长度才会增加
                    left = mid + 1;
                } else {
                    right = mid;
                }
            }
            // 放入递增子序列中, 如果遇到nums中的降序, 这个位置会被覆盖
            tails[left] = num;
            // left已经到了tails的右边界,长度加一
            if (left == max) {
                max++;
            }
        }
        return max;
    }

    /**
     * 超时: 性能问题
     */
    public int lengthOfLIS1(int[] nums) {
        backtrack(new ArrayList<>(), 0, nums);
        return result;
    }

    private void backtrack(List<Integer> list, int index, int[] nums) {
        if (list.size() >= 2 && list.size() > result) {
            result = list.size();
        }
        for (int i = index; i < nums.length; i++) {
            if (list.isEmpty() || list.getLast() < nums[i]) {
                list.add(nums[i]);
                backtrack(list, i + 1, nums);
                list.removeLast();
            }
        }
    }
}
// @lc code=end

File: leetcode/src/main/java/org/luvx/leetcode/java/medium/_300/Solution.go
package main

import (
    "fmt"
    "math"
)

/*
 * @lc app=leetcode.cn id=300 lang=golang
 *
 * [300] 最长递增子序列
 */

// @lc code=start
func lengthOfLIS(nums []int) int {
    dp := make([]int, 0)
    dp = append(dp, nums[0])
    for i := 1; i < len(nums); i++ {
        num := nums[i]
        if num > dp[len(dp)-1] {
            dp = append(dp, num)
            continue
        }
        for j := 0; j < len(dp); j++ {
            if dp[j] >= num {
                dp[j] = num
                break
            }
        }
    }
    return len(dp)
}
func lengthOfLIS0(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    dp := make([]int, len(nums)+1)
    dp[0] = math.MinInt
    dp[1] = nums[0]

    maxL := 1
    for i := 1; i < len(nums); i++ {
        num := nums[i]
        if dp[maxL] < num {
            maxL++
            dp[maxL] = num
            continue
        }

        for j := maxL - 1; j >= 0; j-- {
            if num > dp[j] {
                println(j)
                dp[j+1] = num
                break
            }
        }
        // 使用二叉搜索来确定 num 应放入 dp 数组中的哪个位置
        //j := searchDP(num, dp, maxL)
        //dp[j+1] = num
    }
    return maxL
}

// searchDP 不正确
func searchDP(num int, dp []int, max int) int {
    left, right := 0, max-1
    for left < right {
        mid := left + (right-left)/2
        if dp[mid] < num {
            left = mid
        } else {
            right = mid
        }
    }
    return left
}

// @lc code=end

func main() {
    lis := lengthOfLIS([]int{4, 10, 4, 3, 8, 9})
    fmt.Println("---\n", lis)
}
